[1.13]Showing exception when transfering to another server.

10:04:42 [ERROR] Exception happened while handling incoming packet AvailableCommandsPacket
java.lang.NullPointerException
        at org.itxtech.nemisys.network.protocol.mcpe.AvailableCommandsPacket.lambda$encode$3(AvailableCommandsPacket.java:275)
        at java.util.HashMap.forEach(HashMap.java:1289)
        at org.itxtech.nemisys.network.protocol.mcpe.AvailableCommandsPacket.encode(AvailableCommandsPacket.java:246)
        at org.itxtech.nemisys.Player.sendCommandData(Player.java:664)
        at org.itxtech.nemisys.Player.handleIncomingPacket(Player.java:220)
        at org.itxtech.nemisys.Player.onUpdate(Player.java:291)
        at org.itxtech.nemisys.Server.lambda$tick$2(Server.java:473)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

--- a/src/main/java/org/itxtech/nemisys/Client.java
+++ b/src/main/java/org/itxtech/nemisys/Client.java
@@ -162,8 +162,9 @@ public class Client {
                 Player pl = players.get(uuid);
                 if (pl != null) {
                     byte[] buffer = ((RedirectPacket) packet).mcpeBuffer;
+                    if (buffer.length == 0) break;
                     DataPacket send;
-                    if (buffer.length > 0 && buffer[0] == (byte) 0xfe) {
+                    if (buffer[0] == (byte) 0xfe) {
                         send = new BatchPacket();
                         send.setBuffer(buffer, 1);
                     } else {

--- a/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/LoginPacket.java
+++ b/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/LoginPacket.java
@@ -151,7 +151,8 @@ public class LoginPacket extends DataPacket {
         byte[] data = Base64.getDecoder().decode(element.get("Image").getAsString());
         int width = element.get("ImageWidth").getAsInt();
         int height = element.get("ImageHeight").getAsInt();
-        return new SkinAnimation(new SerializedImage(width, height, data), type, frames);
+        int expression = element.get("AnimationExpression").getAsInt();
+        return new SkinAnimation(new SerializedImage(width, height, data), type, frames, expression);
     }
 
     private static SerializedImage getImage(JsonObject token, String name) {

--- a/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/ProtocolInfo.java
+++ b/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/ProtocolInfo.java
@@ -9,10 +9,10 @@ public interface ProtocolInfo {
     /**
      * Actual Minecraft: PE protocol version
      */
-    int CURRENT_PROTOCOL = Integer.valueOf("419"); //plugins can change it
+    int CURRENT_PROTOCOL = Integer.valueOf("422"); //plugins can change it
 
-    String MINECRAFT_VERSION = "v1.16.100";
-    String MINECRAFT_VERSION_NETWORK = "1.16.100";
+    String MINECRAFT_VERSION = "v1.16.200";
+    String MINECRAFT_VERSION_NETWORK = "1.16.200";
 
     byte LOGIN_PACKET = 0x01;
     byte PLAY_STATUS_PACKET = 0x02;

--- a/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/ProtocolInfo.java
+++ b/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/ProtocolInfo.java
@@ -9,10 +9,10 @@ public interface ProtocolInfo {
     /**
      * Actual Minecraft: PE protocol version
      */
-    int CURRENT_PROTOCOL = Integer.valueOf("408"); //plugins can change it
+    int CURRENT_PROTOCOL = Integer.valueOf("419"); //plugins can change it
 
-    String MINECRAFT_VERSION = "v1.16.20";
-    String MINECRAFT_VERSION_NETWORK = "1.16.20";
+    String MINECRAFT_VERSION = "v1.16.100";
+    String MINECRAFT_VERSION_NETWORK = "1.16.100";
 
     byte LOGIN_PACKET = 0x01;
     byte PLAY_STATUS_PACKET = 0x02;

--- a/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/ProtocolInfo.java
+++ b/src/main/java/org/itxtech/nemisys/network/protocol/mcpe/ProtocolInfo.java
@@ -9,10 +9,10 @@ public interface ProtocolInfo {
     /**
      * Actual Minecraft: PE protocol version
      */
-    int CURRENT_PROTOCOL = Integer.valueOf("407"); //plugins can change it
+    int CURRENT_PROTOCOL = Integer.valueOf("408"); //plugins can change it
 
-    String MINECRAFT_VERSION = "v1.16.0";
-    String MINECRAFT_VERSION_NETWORK = "1.16.0";
+    String MINECRAFT_VERSION = "v1.16.20";
+    String MINECRAFT_VERSION_NETWORK = "1.16.20";
 
     byte LOGIN_PACKET = 0x01;
     byte PLAY_STATUS_PACKET = 0x02;

--- a/src/main/java/org/itxtech/nemisys/utils/BinaryStream.java
+++ b/src/main/java/org/itxtech/nemisys/utils/BinaryStream.java
@@ -225,6 +225,7 @@ public class BinaryStream {
             this.putImage(animation.image);
             this.putLInt(animation.type);
             this.putLFloat(animation.frames);
+            this.putLInt(animation.expression);
         }
 
         this.putImage(skin.getCapeData());
@@ -270,7 +271,8 @@ public class BinaryStream {
             SerializedImage image = this.getImage();
             int type = this.getLInt();
             float frames = this.getLFloat();
-            skin.getAnimations().add(new SkinAnimation(image, type, frames));
+            int expression = this.getLInt();
+            skin.getAnimations().add(new SkinAnimation(image, type, frames, expression));
         }
 
         skin.setCapeData(this.getImage());

--- a/src/main/java/org/itxtech/nemisys/utils/BinaryStream.java
+++ b/src/main/java/org/itxtech/nemisys/utils/BinaryStream.java
@@ -533,6 +533,10 @@ public class BinaryStream {
             }
         }
 
+        if (id == 513) { // TODO: Shields
+            this.getVarLong();
+        }
+
         return new Item(
                 id, data, cnt, nbt
         );
@@ -552,6 +556,10 @@ public class BinaryStream {
         this.put(nbt);
         this.putVarInt(0); //TODO CanPlaceOn entry count
         this.putVarInt(0); //TODO CanDestroy entry count
+
+        if (item.getId() == 513) { // TODO: Shields
+            this.putVarLong(0);
+        }
     }
 
     public BlockVector3 getSignedBlockPosition() {

--- a/src/main/java/org/itxtech/nemisys/utils/SkinAnimation.java
+++ b/src/main/java/org/itxtech/nemisys/utils/SkinAnimation.java
@@ -7,10 +7,12 @@ public class SkinAnimation {
     public final SerializedImage image;
     public final int type;
     public final float frames;
+    public final int expression;
 
-    public SkinAnimation(SerializedImage image, int type, float frames) {
+    public SkinAnimation(SerializedImage image, int type, float frames, int expression) {
         this.image = image;
         this.type = type;
         this.frames = frames;
+        this.expression = expression;
     }
 }

--- a/src/main/java/org/itxtech/nemisys/utils/VarInt.java
+++ b/src/main/java/org/itxtech/nemisys/utils/VarInt.java
@@ -16,7 +16,7 @@ public class VarInt {
      */
     public static long encodeZigZag32(int v) {
         // Note:  the right-shift must be arithmetic
-        return (long) ((v << 1) ^ (v >> 31));
+        return ((v << 1) ^ (v >> 31)) & 0xFFFFFFFFL;
     }
 
     /**

